manual/src/cmds/tail-mod-cons.etex

```diff
@@ -0,0 +1,528 @@
+\begin{caml_example*}{verbatim}
+let rec map f l =
+  match l with
+  | [] -> []
+  | x :: xs ->
+    let y = f x in
+    y :: map f xs
+\end{caml_example*}
+
+A well-known limitation of this implementation is that the recursive
+call, "map f xs", is not in \emph{tail} position. The runtime needs to
+remember to continue with "y :: r" after the call returns a value "r",
+therefore this function consumes some amount of call-stack space on
+each recursive call. The stack usage of "map f li" is proportional to
+the length of "li". This is a correctness issue for large lists on
+operating systems with limited stack space -- the dreaded
+"Stack_overflow" exception.
+
+\begin{caml_example}{toplevel}
+List.length (map Fun.id (List.init 1_000_000 Fun.id));;
+\end{caml_example}
+
+In this implementation of "map", the recursive call happens in
+a position that is not a \emph{tail} position in the program, but
+within a datatype constructor application that is itself in
+\emph{tail} position. We say that those positions, that are composed
+of tail positions and constructor applications, are \emph{tail modulo
+  constructor} (TMC) positions -- we sometimes write \emph{tail modulo
+  cons} for brevity.
+
+It is possible to rewrite programs such that tail modulo cons
+positions become tail positions; after this transformation, the
+implementation of "map" above becomes \emph{tail-recursive}, in the
+sense that it only consumes a constant amount of stack space. The
+OCaml compiler implements this transformation on demand, using the
+"[\@tail_mod_cons]" or "[\@ocaml.tail_mod_cons]" attribute on the
+function to transform.
+
+\begin{caml_example*}{verbatim}
+let[@tail_mod_cons] rec map f l =
+  match l with
+  | [] -> []
+  | x :: xs ->
+    let y = f x in
+    y :: map f xs
+\end{caml_example*}
+
+\begin{caml_example}{toplevel}
+List.length (map Fun.id (List.init 1_000_000 Fun.id));;
+\end{caml_example}
+
+This transformation only improves calls in tail-modulo-cons position,
+it does not improve recursive calls that do not fit in this fragment:
+\begin{caml_example*}{verbatim}[warning=71]
+(* does *not* work: addition is not a data constructor *)
+let[@tail_mod_cons] rec length l =
+  match l with
+  | [] -> 0
+  | _ :: xs -> 1 + length xs
+\end{caml_example*}
+
+It is of course possible to use the "[\@tail_mod_cons]" transformation
+on functions that contain some recursive calls in tail-modulo-cons
+position, and some calls in other, arbitrary positions. Only the tail
+calls and tail-modulo-cons calls will happen in constant stack space.
+
+\paragraph{General design} This feature is provided as an explicit
+program transformation, not an implicit optimization. It is
+annotation-driven: the user is expected to express their intent by
+adding annotations in the program using attributes, and will be asked
+to do so in any ambiguous situation.
+
+We expect it to be used mostly by advanced OCaml users needing to get
```

Comment:
This is exclusive `we` seems out of place when the manual (and even this chapter) generally use an inclusive `we`.